javascript/src/wombatSetup.js

```diff
@@ -26,63 +27,62 @@ export function urlRewriteFunction(
 ) {
   if (!url) return url;
 
-  url = url.toString();
-
-  if (url.startsWith(orig_host)) return url;
-
-  for (const prefix of [
-    '#',
-    'about:',
-    'data:',
-    'blob:',
-    'mailto:',
-    'javascript:',
-    '{',
-    '*',
-  ]) {
-    if (url.startsWith(prefix)) {
-      return url;
-    }
-  }
-
-  var absolute_url;
-  if (url.startsWith('//')) {
-    absolute_url = orig_scheme + url;
-  } else if (url.startsWith('/')) {
-    // We have a absolute path without host.
-    // So it is a absolute path relative to the original host.
-    absolute_url = new URL(url, orig_url).toString();
-  } else {
-    // Relative path or full url.
-    // Let's build relative to our current url (`URL` will take care of relative vs full url)
-    absolute_url = new URL(url, current_url).toString();
-  }
-
-  var entry_path;
-  if (absolute_url.startsWith(prefix)) {
-    // The absolute_url start with our prefix.
-    // It means that `url` was a relative or already processed url.
-    // We can simply remove our prefix to found our entry's path.
-    entry_path = absolute_url.substring(prefix.length);
-  } else {
-    // Remove potential scheme.
-    entry_path = absolute_url.replace(/^\w+:?\/\//i, '');
-  }
-
-  // Now we have a entry's path "as seen by the website".
-  // We need to reduce it to what we have stored in the zim file.
-  const reduced_path = applyFuzzyRules(entry_path);
-
-  var final_url = prefix + reduced_path;
-
-  final_url = new URL(final_url);
-  if (final_url.search) {
-    final_url.pathname =
-      final_url.pathname + encodeURIComponent(final_url.search);
-    final_url.search = '';
+  // Special stuff which is not really a URI but exists in the wild
+  if (['#', '{', '*'].includes(url.substring(0, 1))) return url;
+
+  // This is a hack to detect improper URL encoding ; proper detection should be
+  // possible with chardet or other alternatives but did not worked so far ; we hence
+  // take benefit of the error below to detect improper URL encoding
+  // When improper URL encoding is detected, we try to encode URL as a best-effort;
+  // 'best-effort', because if some part of the URL is encoded and another part is not,
+  // this will fail ... but this is a weird edge case anyway
+  try {
+    decodeURIComponent(URI.parse(url).path);
+  } catch (e) {
+    url = encodeURI(url);
   }
 
-  return final_url.toString();
+  // If URI scheme is defined but not http or https, we have to not rewrite the URL
+  const uri = URI.parse(url);
+  if (
+    typeof uri.scheme !== 'undefined' &&
+    !['http', 'https'].includes(uri.scheme)
+  )
+    return url;
+
+  // Compute the absolute URI, just like the browser would have resolved it hopefully
+  const original_absolute_url = URI.resolve(orig_url, url);
```

Comment:
Initial code resolve relative url based on `current_url`, not `orig_url`.
This is important as current url is not the same as orig url if fuzzy matching has been applied.

Reply:
@benoit74 was this intentional? If yes why and comment

Reply:
This is intentional, but I agree there is still a bug. I'm looking into it right now. 

Using `orig_url` is mandatory to properly rewrite the URL when it is an absolute URL on a "new" hostname and we want to properly encode the hostname. Maybe the trick is to make a difference between the case where we have an absolute URL and the case where it is a relative one (which might hence be already rewritten, we just don't know).

Reply:
I will try to summarize the bug (I'm not sure yet on how to solve it).

Let's say:
- the ZIM is served at `https://mylibrary.com/content/myzim/`
- we are on a page originally located at `https://www.example.com/index.html`
- Wombat passes us the following URL to rewrite: `../www.otherdomain.com/image.png`. 

I don't know how to differentiate two cases with 100% certainty.

Case 1: 
- the URL has not been rewritten yet 
- the resource was originally located at `https://www.example.com/www.otherdomain.com/image.png`
- the rewritten URL must be `https://mylibrary.com/content/myzim/www.example.com/www.otherdomain.com/image.png`
- there was a silent bug in HTML code where the relative link was going "too high" in the tree, but AFAIK most browser are very permissive with this scenario

Case 2:
- the URL has already been rewritten
- the resource was originally located at `https://www.otherdomain.com/image.png`
- the rewritten URL must be `https://mylibrary.com/content/myzim/www.otherdomain.com/image.png`

A dirty hack would be to include a magic querystring (or prefix, or ...) in all rewritten URL so that we are 100% percent sure in which scenario we are.

Another solution would be to use a heuristic: if we are rewriting a relative link AND there is at least one dot in the first path segment (indicating a probably hostname, quite rare in directories) AND the link is going exactly one level too high, then we can be almost 100% confident the URL to rewrite has indeed already been rewritten and we can keep it as-is.

As you can see, using `current_url` instead of `orig_url` does not help at all, it just assumes that we are always in case 2 by allowing JS to go one level higher when resolving the `../`.

I would prefer to implement the heuristic, but did I missed a more straightforward solution or am I wrong somewhere? Any idea?

Reply:
We've discussed it live with @mgautierfr 

We've chosen a middle ground:
- rewrite all domain names by appending a `.zim` suffix 
- use that to detect that a relative URL has already been rewritten
- document that domains names ending with `.zim` are not supported (this TLD does not yet exists)

Reply:
Edit: rewriting all domain names by appending a `.zim` suffix does not helped at all indeed. We still need to detect the difference between a real path ending with `.zim` and a "fake / ZIM-only" path ending with `.zim` because it is the domain which has been encoded there.

I hence finally implemented the heuristic I spoke about in previous comment, implemented a test-case on the test website and everything is working fine now (while it was broken before).